### PR TITLE
Optimize parse method

### PR DESCRIPTION
**Pull request by jhthorsen**
This PR makes the parse() method faster.

I got a [benchmark test](https://github.com/jhthorsen/mojo-redis/blob/master/t/benchmark.t) that I've used to check how much faster this version is with the command below:

```
TEST_BENCHMARK=1000 TEST_ONLINE=redis://localhost prove -vl t/benchmark.t
```

## With the code in this PR

```
ok 1 - lrange Protocol::Redis 2.707 wallclock secs ( 2.54 usr +  0.05 sys =  2.59 CPU) @ 386.10/s (n=1000)
ok 2 - lrange Protocol::Redis::XS 1.84939 wallclock secs ( 1.69 usr +  0.06 sys =  1.75 CPU) @ 571.43/s (n=1000)
# Cannot compare Redis/PP and Redis2/PP
                     Rate     Protocol::Redis Protocol::Redis::XS
Protocol::Redis     386/s                  --                -32%
Protocol::Redis::XS 571/s                 48%                  --
```

## Running against 1.0006

```
ok 1 - lrange Protocol::Redis 6.00857 wallclock secs ( 5.81 usr +  0.07 sys =  5.88 CPU) @ 170.07/s (n=1000)
ok 2 - lrange Protocol::Redis::XS 1.84052 wallclock secs ( 1.68 usr +  0.05 sys =  1.73 CPU) @ 578.03/s (n=1000)
# Cannot compare Redis/PP and Redis2/PP
                     Rate     Protocol::Redis Protocol::Redis::XS
Protocol::Redis     170/s                  --                -71%
Protocol::Redis::XS 578/s                240%                  --
```

### CODE REVIEW

**Comment by jhthorsen** [comment]
I forced pushed to fix some typos and improve the performance further.

File: lib/Protocol/Redis.pm [hunk 1]
```diff
@@ -91,194 +89,89 @@ sub get_message {
...
         }
-        else {
 
-            # read next string
-            $self->{_state_cb} = $mbulk_process;
-            $self->{_state}    = \&_state_parse_message_type;
-            $chunk;
+        # Should not come to this
+        else {
+            Carp::croak(qq/Unknown message type $curr->{type}/);
```
Comment (Grinnz) [inline lib/Protocol/Redis.pm]: This should maintain the existing error message: `Unexpected input "$cmd"`

**Comment by und3f** [comment]
Please excuse my silence, I've had some issues. I've starting reviewing this merge request.
Also it is great that you've added P:R:Fast!

**Comment by und3f** [comment]
Well, the code is complex, and not so easy to read, will spend some time on it

**Comment by und3f** [comment]
Thank you for the commit, now it is merged. Also I've got some optimizations. According to my results now P:R parse() method is faster than P:R:F

````
                Rate   P:R encode P:R:F encode
P:R encode   52852/s           --          -0%       
P:R:F encode 53066/s           0%           --  
                       Rate P:R:F Status parse   P:R Status parse
P:R:F Status parse 321555/s                 --               -13%
P:R Status parse   367589/s                14%                 --
                     Rate Bulk parse P:R:F   Bulk parse P:R                
Bulk parse P:R:F 157538/s               --             -12%
Bulk parse P:R   178641/s              13%               --
                           Rate P:R:F Multi-Bulk parse   P:R Multi-Bulk parse
P:R:F Multi-Bulk parse  95999/s                     --                    -6%
P:R Multi-Bulk parse   102399/s                     7%                     --

**Comment by und3f** [comment]
> One more thing, from [Grinnz/Protocol-Redis-Faster@0df649e](https://github.com/Grinnz/Protocol-Redis-Faster/commit/0df649e9148f6b38955dec2a9e645c42b14a7608), $self->{_message} should be deleted before calling on_message_cb in case it calls parse itself.

thanks! Fixed in 2d13250eb7c8bcd1cc504fe4fd360246d894fff0

**Pull request merged**